### ISSUE TITLE
Redirect rule is not triggered for unpublished product

### ISSUE THREAD

**Issue by lem8r** [odoo/odoo#64778]
Impacted versions: 14.0, 13.0

Steps to reproduce: Unpublish product page, create 301 redirect for that page. Visit unpublished page as public or portal user.

Current behavior: 403 Access error

Expected behavior: 301 redirect

Same will happen for any router with model when record is unpublished.
Redirect rule [is checked](https://github.com/odoo/odoo/blob/084d06a5892ab2723d02dc29f0e62377c554efac/odoo/addons/base/models/ir_http.py#L191) only for exceptions with 404 code but security rule `website_page_rule_public` raises generic `AccessError` exception.

**Comment by Yenthe666** [comment]
@JKE-be

**Comment by JKE-be** [comment]
Yes it is by design from this way:https://github.com/odoo/odoo/blob/13.0/odoo/addons/base/models/ir_http.py#L191

We check if it match a controller (controller are always priority 1)

- If no controller found (whatever the status):
   we check if attachment with url

- if not attachment 
  we check if a page

- if not page
  we check if a redirect

As soon once is found: controller -> attachment -> page -> redirect... We don't try to find a fallback (see def _serve_fallback)

Maybe an future improvement for master.

**Comment by lem8r** [comment]
It is common requirement to have redirects for unpublished products as link is still posted on social media, cached by crawlers, etc.

**Comment by JKE-be** [comment]
Yes, I'm totally agree with you.
I created a task internally to improve it in master.

But it is not really considered as bug because it was implemented to work from this way. And We cannot change it in stable because people already work with it for now.

(Perso, I myself customize my 403 to redirect with meta to a x_alternate_product, not clean... but an example of custo that will be broken if we change it in stable)
